### PR TITLE
Simplify Encryptor API Surface

### LINKED ISSUE #843: Move salt to encryptor.js

### LINKED ISSUE #859: Finalize encryptor.js serialized format
I'm not happy with the current encryptor serialized format, I'd like to make it more explicit for someone just reading it, maybe making it a JSON payload where the portions of the payload are labeled correctly.

Maybe: `{ data, iv, salt }`, where `iv` is the vector argument named `iv` in the `SubtleCrypto` browser API?

`data` and `iv` would be `base64`, but `salt` would be a string to be appended to the vault's password.

### PR DESCRIPTION
**Pull request by danfinlay**
I'd like to get honest opinions about this change before we accept & merge it.

This simplifies our code a bit, but has a performance tradeoff.

In this branch, KeyringController caches the password instead of a passwordDerivedKey, since it is ignorant of the salt, which means more crypto each time we need to persist the vault (re-deriving the key), but simplifies the code in KeyringController.

This was necessary because we moved salting within the encryptor, so the salt does not need to be managed externally.

Encryptor payload is now in a JSON format, so its portions are both base64 encoded *and* labeled appropriately.  The format is `{ "data": "base64", "iv": "base64", "salt": "string" }`.

Fixes #843
Fixes #859

### CODE REVIEW

File: app/scripts/keyring-controller.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ module.exports = class KeyringController extends EventEmitter {
     return {
       seedWords: this.configManager.getSeedWords(),
       isInitialized: (!!wallet || !!vault),
-      isUnlocked: !!this.key,
+      isUnlocked: Boolean(this.password),
```
Comment (kumavis) [inline app/scripts/keyring-controller.js]: 👍

File: app/scripts/keyring-controller.js [hunk 2]
```diff
@@ -113,34 +113,25 @@ module.exports = class KeyringController extends EventEmitter {
...
-    })
+    return this.idStoreMigrator.migratedVaultForPassword(password)
     .then((serialized) => {
       if (serialized && shouldMigrate) {
         const keyring = this.restoreKeyring(serialized)
         this.keyrings.push(keyring)
         this.configManager.setSelectedAccount(keyring.getAccounts()[0])
         return this.persistAllKeyrings()
-        .then(() => { return key })
+        .then(() => { return })
```
Comment (kumavis) [inline app/scripts/keyring-controller.js]: do we need this?

**Review by kumavis — APPROVED** [review]

File: app/scripts/keyring-controller.js (hunk 2 above)
Comment (flyswatter) [inline app/scripts/keyring-controller.js]: Oh yeah probably not

File: app/scripts/keyring-controller.js (hunk 1 above)
Comment (flyswatter) [inline app/scripts/keyring-controller.js]: I didn't really know about this method before today when I saw you do it.

File: app/scripts/keyring-controller.js [hunk 3]
```diff
@@ -285,16 +265,16 @@ module.exports = class KeyringController extends EventEmitter {
...
       this.configManager.setVault(encryptedString)
       return true
     })
   }
 
-  unlockKeyrings (key) {
+  unlockKeyrings (password) {
     const encryptedVault = this.configManager.getVault()
-    return this.encryptor.decryptWithKey(key, encryptedVault)
+    return this.encryptor.decrypt(this.password, encryptedVault)
```
Comment (kumavis) [inline app/scripts/keyring-controller.js]: this looks wrong `password` vs `this. password`

**Pull request merged**